rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/LocalInfrastructure.java

```diff
@@ -333,34 +356,34 @@ protected void configure(Object... args) {
      */
     @Override
     protected void notifyDeployingNodeLost(String pnURL) {
-        this.lostNodes.incrementAndGet();
+        incrementLostNodes();
     }
 
     @Override
     protected void notifyAcquiredNode(Node arg0) throws RMException {
-        this.acquiredNodes.incrementAndGet();
+        incrementAcquiredNodes();
     }
 
     @Override
     public void removeNode(Node node) throws RMException {
-        logger.debug("Removing node " + node.getNodeInformation().getURL() + " from " +
+        logger.debug("A node is removed " + node.getNodeInformation().getURL() + " from " +
                      this.getClass().getSimpleName());
 
-        if (!this.nodeSource.getDownNodes().contains(node)) {
-            // the node was manually removed
-            handledNodes.decrementAndGet();
-        }
+        // the node was removed intentionally, so we decrement the number of handled nodes
+        decrementHandledNodes();
+        // and the number of acquired nodes is decremented too
+        removeAcquiredNodesAndShutDownIfNeeded();
+    }
 
-        int remainingNodesCount = this.acquiredNodes.decrementAndGet();
-        // If there is no remaining node, kill the JVM process
-        if (remainingNodesCount == 0) {
-            shutDown();
-        }
+    @Override
+    public void notifyDownNode(String nodeName, String nodeUrl, Node node) {
```

Comment:
Why is this called notifyDownNode? Isn't it removing a down node? Maybe it is possible to change the underlying interface for better naming?

Reply:
Basically, this method represent what the infrastructure implementation needs to do when a node is down.
As such, this is a hook that I have named the same way as the other hooks for different events, like `notifyDeployingNodeLost` and `notifyAcquiredNode`. I agree that I would rather name the new hook `onDownNode` but in this case I would rename also the methods `onDeployingNodeLost`and `onAquiredNode`.

Reply:
The method is a notification to the infrastructure. The RM notifies the infrastructure that a node is down. Then it is up to the infrastructure to decide what to do with it.